### PR TITLE
Add -Wconversion to various things in src/app

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
We are not using `-Wconversion` to build things in `src/app`, so we won't catch problems as we start changing the sizes of things like cluster/endpoint/attribute ids.

#### Change overview
Add `-Wconversion` to some of the relevant .gn/.gni files, fix resulting problems.

#### Testing
Compiled the code on esp32, Mac, nrfconnect.

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

File: src/app/util/attribute-storage.cpp [hunk 1]
```diff
@@ -176,13 +176,15 @@ uint8_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
 EmberAfStatus emberAfSetDynamicEndpoint(uint8_t index, EndpointId id, EmberAfEndpointType * ep, uint16_t deviceId,
                                         uint8_t deviceVersion)
 {
-    index += FIXED_ENDPOINT_COUNT;
+    auto realIndex = index + FIXED_ENDPOINT_COUNT;
```
Comment (mspang) [inline src/app/util/attribute-storage.cpp:179]: auto -> int?

**Review by mspang — APPROVED** [review]

File: src/app/util/attribute-storage.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/util/attribute-storage.cpp:179]: I could do that, but the point is we don't really care what this gets promoted to by C's rules.  We just check that it's in-range and then cast it to the type we really want...  So I was pretty carefully avoiding using `int` there.

Comment (mspang) [reply src/app/util/attribute-storage.cpp:179]: It's not that we care or don't care. It's that the type is clear and it's "int". It isn't going to change and there aren't any other promotion rules like that. Why obfuscate?

**Comment by bzbarsky-apple** [comment]
@Damian-Nordic @jmartinez-silabs Please take a look?

**Review by woody-apple — APPROVED** [review]

**Pull request merged**